erpnext/stock/doctype/stock_entry/stock_entry.py

```diff
@@ -918,133 +922,210 @@ def set_items_for_stock_in(self):
-					self.purpose == "Material Consumption for Manufacture") and self.flags.backflush_based_on== "BOM"
-					and frappe.db.get_single_value("Manufacturing Settings", "material_consumption")== 1):
-					self.get_unconsumed_raw_materials()
+			if self.get("wo_doc"):
+				add_additional_cost(self, self.wo_doc)
 
-				else:
-					if not self.fg_completed_qty:
-						frappe.throw(_("Manufacturing Quantity is mandatory"))
-
-					item_dict = self.get_bom_raw_materials(self.fg_completed_qty)
-
-					#Get PO Supplied Items Details
-					if self.purchase_order and self.purpose == "Send to Subcontractor":
-						#Get PO Supplied Items Details
-						item_wh = frappe._dict(frappe.db.sql("""
-							select rm_item_code, reserve_warehouse
-							from `tabPurchase Order` po, `tabPurchase Order Item Supplied` poitemsup
-							where po.name = poitemsup.parent
-								and po.name = %s""",self.purchase_order))
-
-					for item in itervalues(item_dict):
-						if self.pro_doc and cint(self.pro_doc.from_wip_warehouse):
-							item["from_warehouse"] = self.pro_doc.wip_warehouse
-						#Get Reserve Warehouse from PO
-						if self.purchase_order and self.purpose=="Send to Subcontractor":
-							item["from_warehouse"] = item_wh.get(item.item_code)
-						item["to_warehouse"] = self.to_warehouse if self.purpose=="Send to Subcontractor" else ""
-
-					self.add_to_stock_entry_detail(item_dict)
-
-			# fetch the serial_no of the first stock entry for the second stock entry
-			if self.work_order and self.purpose == "Manufacture":
-				self.set_serial_nos(self.work_order)
-				work_order = frappe.get_doc('Work Order', self.work_order)
-				add_additional_cost(self, work_order)
-
-			# add finished goods item
-			if self.purpose in ("Manufacture", "Repack"):
-				self.load_items_from_bom()
-
-		self.set_scrap_items()
-		self.set_actual_qty()
 		self.calculate_rate_and_amount(raise_error_if_no_rate=False)
 
-	def set_scrap_items(self):
-		if self.purpose != "Send to Subcontractor" and self.purpose in ["Manufacture", "Repack"]:
-			scrap_item_dict = self.get_bom_scrap_material(self.fg_completed_qty)
-			for item in itervalues(scrap_item_dict):
-				item.idx = ''
-				if self.pro_doc and self.pro_doc.scrap_warehouse:
-					item["to_warehouse"] = self.pro_doc.scrap_warehouse
+	def set_items_from_work_order(self):
+		remain_qty_to_produce = self.get_remain_qty_to_produce()
+
+		for row in self.wo_doc.required_items:
+			if (self.purpose == "Material Transfer for Manufacture"
+				and (self.wo_doc.skip_transfer or row.skip_material_transfer)): continue
+
+			if (row.transferred_qty and row.transferred_qty >= row.required_qty
+				and row.consumed_qty >= row.transferred_qty):
+				continue
+
+			transferred_qty = row.required_qty
+			if row.transferred_qty > row.required_qty:
+				transferred_qty = row.transferred_qty
+
+			qty = ((transferred_qty - row.consumed_qty)
+				* flt(self.fg_completed_qty)) / remain_qty_to_produce
```

Comment:
```suggestion
			qty = (flt(transferred_qty - row.consumed_qty)
				* flt(self.fg_completed_qty)) / (remain_qty_to_produce or 1)
```
denominator could be 0 if not is_return.